### ISSUE TITLE
Play VGLC MegaMan Levels?

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#473]
Before getting too deep into #464  I would like to check if we can actually play any of the levels we create. This would require some kind of source code for a MegaMan clone, preferably in Java, though we could probably work with Python as well. I've found several candidates, but all are a bit lacking in quality. The thing we really need is support for a platforming game that includes shooting and ladders. Anything else is just icing on the cake. There also needs to be a way of mapping the VGLC levels into the format of the code we use (that would be the main work required of this issue).

Anyway, here are links to the candidates code I found:
https://github.com/omega87910/106-2-Project-Rockman (constant scroll ... no platforms/ladders)
https://github.com/R-Wang97/Megaman (Has platforms, though the movement mechanics are weird)
http://pesquisa.ufabc.edu.br/hal/Evoman.html (in Python. Mainly meant for boss fights ... not sure what level editing options there are)

We may need to find others or write some code ourselves.

**Comment by schrum2** [comment]
I'm nor sure if any of the above code repos are really worth using, but an alternative occurred to me. We don't have to have the source code (or at least, we don't need to mess with the source code) for Mega Man. All we really need is a level editor, and some way of launching levels and getting back a result. So, we might be able to use a pure executable, ROM hack, or C++ version of the game. Here is what I've found (haven't looked too closely yet):

C++ remake of the game:
https://github.com/hakase-labs/hikari
Might be difficult to compile, but the level format seems to be json, which is easy to manipulate.

ROM level editor:
https://bisqwit.iki.fi/source/rockedit.html
I'm not sure how complete/functioning this is, but it looks like you can specify a Mega Man level with a text file format similar to what VGLC has, and patch the original Mega Man ROM with it. This might be too slow to be practical, but is worth considering.

**Comment by schrum2** [comment]
Played the latest commit: Allow scrolling to the left as well. Additionally, when moving/scrolling, the enemy movement is also affected. Last, center Mega Man in the middle of the screen instead of the far left.

**Comment by schrum2** [comment]
Mega Man Maker ( https://megamanmaker.com/ ) seems like a promising option to pursue. The Java code for Project Rockman is probably too incomplete to really work with. Although this means the work in the branch will likely be discarded, the prospect of using Mega Man Maker is appealing.

In order to figure out if Mega Man Maker (MMM) is actually a viable option, we have to make sense of the level text representation. If you can convert a VGLC level into MMM format, then this issue can be considered finished/closed @cappsb

**Comment by cappsb** [comment]
Here's a link to the source code that edits levels in the MMLV format. If we read it carefully, then we may be able to make sense of how it works.
https://github.com/godot-mega-man/mega-man-maker-mmlv-editor/blob/master/Project/src/database/game_bg_data.gd


Some user insight as to what numbers do what:
https://megamanmaker.com/forum/viewtopic.php?t=2371
Basically:
Every single-letter at the top represents a part of what the block does (it may be solid, connected to another, it may have a timer associated with it)
Each pair of numbers is the coordinate
The number in "quotes" represents the individual quality
things beginning in 2d are the background images
final numbers are the version number, name of user, etc.

**Comment by cappsb** [comment]
Additional information as to the format (more clear than forum link)
https://gist.github.com/freem/03d6c1832299e8d8a32f64361380920d

**Comment by cappsb** [comment]
Able to generate SOME of the level, but it stops producing squares after a little while
![Screenshot (227)](https://user-images.githubusercontent.com/63438571/83556539-b3d08d00-a4d5-11ea-9ad0-e519b3fe49ce.png)
![Screenshot (229)](https://user-images.githubusercontent.com/63438571/83556560-b9c66e00-a4d5-11ea-827b-46c57ba56dcd.png)

Found out what the problem was:
The variables 1s and 1p correspond to the height and width of the number of blocks being placed. If the number is not large enough, then it simply does not spawn. If the number is too big, it throws an error and crashes. Additionally, I found that the original MegaMan screen is only 15 blocks tall, whereas the editor has 16 blocks per screen. As such, I had to go back and change all of the screens to be scrollers and not fast-transitioning. The problem with the fast transitioning was that the level was cut in half vertically and whenever you jump, the screen jumps up to the next screen. When you fall, the screen stops and moves back down.
![Screenshot (230)](https://user-images.githubusercontent.com/63438571/83556571-bfbc4f00-a4d5-11ea-8c82-b5dd32f9fe6d.png)
Above is a picture of a small portion (can't zoom out) of the entire level spawned (only ladders, spikes, MegaMan, and blocks are included, though more items are easy to add now that I know the formula)

The major problem remaining:
the entire area of the level can be traversed through (it should be null space somewhere the screen would stop instead of being able to continue traversal). I'm working on a solution, but for now, the level is loaded and playable.

**NOTE**
if you're trying to run the commits on your computer, go to around line 163 of MegaManVGLCUtil and change "dakuchen" to whatever your username is.

**Comment by cappsb** [comment]
![image](https://user-images.githubusercontent.com/63438571/83564517-7292aa00-a4e2-11ea-8131-47eaeffdf60d.png)
The problem is minor but unfixable (as far as I know). Level 3 on VGLC is too tall for the builder to handle. When I try to manually extend it in the .mmlv file, the game crashes

**Comment by cappsb** [comment]
Note that some levels are unable to be traversed because they lack moving platforms and breakable objects

**Comment by schrum2** [comment]
Couple of comments:

- Just played your conversion of the first level. Very cool, but I think we should be able to eventually fix the problem of falling off the bottom edge of the viewable screen into a vast emptiness. Eventually, I would like to fix the camera to the appropriate region, and also scroll between sections appropriately. I think that the presence of the null characters in the VGLC format should provide us with a way of doing this.

- For Level 3, which is too tall, are we sure this is a hard limit in MMM, or is there maybe a configuration setting to solve this? Ask on the MMM forum.

- Regarding movable platforms and breakable objects: I believe that MMM supports these, so I suspect the problem is that the VGLC does not provide enough information for us to know which blocks/tiles should be movable/breakable. Is this correct? For levels that have this problem, see if you an manually fix them to behave correctly when played (you might even edit them in the MMM editor and see if you can fix them that way). First, I just want to know what is possible. The next step will be to create our own modified/improved version of the VGLC data (don't overwrite original ... make new subdir for "enhanced" versions) from which the relevant info about movable/breakable blocks can be extracted directly. We might even add enemies to this format eventually.

- Despite all of the problems with the conversion, I would like to have these MMLV files saved to the MMNEAT repo. Make a new subdir in the `data` dir: `data/MegaManMaker/ConvertedVGLCLevels`. Place all converted levels in this location. Over time, as our ability to convert the levels improves, we will overwrite the old versions with new/better ones.

**Comment by cappsb** [comment]
VGLC does in fact support breakable objects and moving platforms. I have not yet implemented them because they are a different size than normal blocks/spikes/ladders, so extra care will need to be taken to place them properly.

**Comment by cappsb** [comment]
Adding breakable blocks works now (see details on how it's done in the commit)
![image](https://user-images.githubusercontent.com/63438571/83573499-374ba780-a4f1-11ea-9be4-a0cb00e92bf2.png)
Still need to do the same process for moving platforms

**Comment by cappsb** [comment]
The map grid on the MMM cannot support the tallest level (level 3)
![Screenshot (231)](https://user-images.githubusercontent.com/63438571/83676999-235f7e80-a5a1-11ea-8605-5ed13c86ef47.png)

Possible solution:
Use % to take y values out of range and spawn the room far off to the right with a teleporter to the bottom of the level that was cut off. This way, when playing, it will be impossible to tell the difference.

**Comment by schrum2** [comment]
How is the player supposed to get past this part of level 10?
![image](https://user-images.githubusercontent.com/8397236/83681629-4b061500-a5a8-11ea-858e-3614e741e93e.png)
Is there an error in the VGLC data, or is there a feature in the level that we are not yet properly representing?

**Comment by schrum2** [comment]
I can't get past this part of Level 6
![image](https://user-images.githubusercontent.com/8397236/83683452-0f207f00-a5ab-11ea-8476-c2f3c027aa27.png)
I know that these platforms are supposed to appear and disappear in a sequence. However, in it's current form I am still able to reach one of the top blocks. The problem is that I can't reach the top right block ... only the one to the left of it. But my head hits the ceiling when I try to jump to it. Even if blocks were disappearing (and thus not being obstacles) I'm not sure how I would make this jump.

**Comment by schrum2** [comment]
In Level 7, I don't have the special weapon needed to break through these blocks:
![image](https://user-images.githubusercontent.com/8397236/83683783-8bb35d80-a5ab-11ea-9621-a980ce6ab0e9.png)
We should make it so that any level that has breakable blocks also lets the player use the weapon able to smash them.

**Comment by schrum2** [comment]
Not really sure where end point is on Level 8. I eventually reach this part:
![image](https://user-images.githubusercontent.com/8397236/83684223-48a5ba00-a5ac-11ea-8a08-125da020add8.png)
But if I go down that final ladder, I just fall and die. Maybe that led to a boss before, but it isn't obvious

**Comment by cappsb** [comment]
1st:
that part of level 10 is the end

2nd:
in the original game, the block (that you are currently hitting your head on) above you disappears long enough for you to jump, then respawns for you to land.

3rd:
I forgot about that, I'll add in the necessary weapons tomorrow

4th:
that's the end

**Comment by schrum2** [comment]
I would like for there to be a main method in whatever util class you think is appropriate that does the following: Loops through all enhanced VGLC text files you have made, and saves the resulting mmlv files into their appropriate subdir in the data directory. The point here is that while you should modify the enhanced VGLC text as much as possible (while leaving original VGLC alone), you should never be manually editing the mmlv files (with a text editor or the MMM program). We want the mmlv files to be completely generated by running your Java code.

**Comment by schrum2** [comment]
I love how level 3 scrolls now. However, I did notice that at one section where there are two ladders, if I use the ladder on the right, the screen scrolls up AND to the right, showing the null space on the side.
![image](https://user-images.githubusercontent.com/8397236/83793693-8a913780-a662-11ea-9813-7fb121d516ac.png)
This is a minor issue, but I wanted to document it so we can fix it eventually.

**Comment by cappsb** [comment]
To resolve the above issue, I looked back at the VGLC data. Turns out, there was "air" right beneath the screen to the right, causing what should be a null space to be completely filled with air. Null placing algorithm tweaked and VGLC data corrected.

**Comment by schrum2** [comment]
A lot of progress has been made on this issue. It mostly works, but we'll keep it open to encourage more fidelity in translating elements from the original game into MegaManMaker. In particular, let's add enemies and bosses, and whatever other mechanics we can.

**Comment by schrum2** [comment]
There are two spots in level 6 where the water placement is off slightly:
![image](https://user-images.githubusercontent.com/8397236/84331211-3cdb6a00-ab4f-11ea-8054-a4efb4f26941.png)
Should be an easy fix.

**Comment by schrum2** [comment]
Level 6 also needs the Level ending orb

**Comment by schrum2** [comment]
Does level 7 still need to have the Level ending orb placed so early, now that you have added the moving enemy platforms over the hazards?

**Comment by schrum2** [comment]
This section of level 9 is impassable:
![image](https://user-images.githubusercontent.com/8397236/84331760-b3c53280-ab50-11ea-91ee-37e9ec380514.png)

It should scroll to the right, but does not.

**Comment by cappsb** [comment]
I've fixed levels six and nine. For level seven, I think we're going to have to keep it there (or move it up a room) because this room has no way to jump up (will still need the platform gun). I checked that level on the original game and sure enough, you're supposed to use your platform gun. Did you want me to add some jumping blocks?

![image](https://user-images.githubusercontent.com/63438571/84395013-4b636900-abc3-11ea-8851-eb2df3101d05.png)
